kernel/src/syscall/unshare.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use ostd::sync::RwArc;
+
+use crate::{namespace::CLONE_NS_FLAGS, prelude::*, process::CloneFlags, syscall::SyscallReturn};
+
+pub fn sys_unshare(unshare_flags: u32, ctx: &Context) -> Result<SyscallReturn> {
+    let flags = CloneFlags::from_bits(unshare_flags)
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "invalid unshare flags"))?;
+
+    debug!("unshare flags = {:?}", flags);
+
+    check_flags(flags, ctx)?;
+
+    unshare_files(flags, ctx);
+    unshare_fs(flags, ctx);
+    unshare_sysvsem(flags, ctx);
+    unshare_namespaces(flags, ctx)?;
+
+    Ok(SyscallReturn::Return(0))
+}
+
+fn check_flags(flags: CloneFlags, ctx: &Context) -> Result<()> {
+    const VALID_FLAGS: CloneFlags = CLONE_NS_FLAGS
+        .union(CloneFlags::CLONE_FILES)
+        .union(CloneFlags::CLONE_FS)
+        .union(CloneFlags::CLONE_SYSVSEM)
+        .union(CloneFlags::CLONE_THREAD)
+        .union(CloneFlags::CLONE_VM)
+        .union(CloneFlags::CLONE_SIGHAND);
+
+    let invalid_flags = flags - VALID_FLAGS;
+    if !invalid_flags.is_empty() {
+        debug!("invalid flags: {:?}", invalid_flags);
+        return_errno_with_message!(Errno::EINVAL, "unsupported unshare flags");
+    }
+
+    if flags.intersects(CloneFlags::CLONE_THREAD | CloneFlags::CLONE_VM | CloneFlags::CLONE_SIGHAND)
+    {
+        let is_single_threaded = ctx.process.tasks().lock().as_slice().len() == 1;
```

Comment:
Please double-check if we want to check this or check the reference count of `Vmar`.

Reply:
Yes. This sounds reasonable. Thanks for the confirmation.